### DISCUSSION TITLE
Help understanding `spin` property of AxisInfo

### DISCUSSION THREAD

**Discussion by jffaust** [cosinekitty/astronomy#266] [Q&A] [answered]
I would like some help understanding the `spin` property of [AxisInfo](https://github.com/cosinekitty/astronomy/tree/master/source/js#axisinfo). I've read the report of the IAU working group linked in the documentation but I'm still a bit lost.

I am trying to make a relatively accurate 3D simulation of the Earth and the Sun in ThreeJS. I'm stuck at applying the correct rotation to my sphere representing the Earth so that it shows the correct area receiving sunlight at a given time.

Here's some additional details about my simulation:
- The Sun is at the origin of my scene and the Earth orbits around it
- I use ecliptic coordinates so that the orbit is horizontal in the scene
- When I apply the Earth texture to my Earth sphere, it is applied horizontally so that the initial north pole is a vector pointing up. Thus, I need to rotate the sphere so that it matches the tilt of the Earth / `north` vector of `AxisInfo` (I also transform this vector in ecliptic coordinates). I use two axes of an Euler rotation to achieve this. The last axis of the rotation lets me make the Earth spin along the adjusted north pole. So far so good.

Now the problem is that I need to calculate a certain spin offset based on the initial position/direction of the prime meridian that is determined by how the texture is applied. 

The IAU report includes the following illustration:

![image](https://user-images.githubusercontent.com/4367222/202859036-19d3e504-e8e0-4b47-9ac1-53eab2a69d15.png)

And we've got the following description: 
> In Fig. 1, the node Q is defined as the node at α0 +90◦. The prime meridian is defined so that it crosses the body’s equator at the point B. The location of B is determined by the value for W, the angle measured easterly along the body’s equator from the Q to B.

From what I undestand, the gamma node or Earth equinox at J2000.0 is the intersection of the Earth's equatorial and ecliptic planes at epoch J2000.0. More precisely, it is the intersection of the March or vernal equinox.

I understand that the ICRF is a reference frame based on the Earth's equator and equinox that has its origin at the barycenter of the solar system at J2000.0. It is almost identical to the [J2000 Frame](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/Tutorials/pdf/individual_docs/04_concepts.pdf): 

![image](https://user-images.githubusercontent.com/4367222/202858295-737f5ed6-1405-4355-a675-3dcee368b707.png)

But I still don't get why Q is gamma plus 90 degrees minus alpha zero.

**Comment by cosinekitty** [comment, marked as answer]
I agree with you, it doesn't make sense. The diagram looks wrong to me; it doesn't match the text, nor does it make sense intuitively. The body's north pole has right ascension $\alpha_0$. That same right ascension is where the body's equator passes most north of the Earth's J2000 equator. The nodes therefore must be at $\alpha_0\pm90^\circ$. The text says we arbitrarily pick $Q=\alpha_0+90^\circ$. The diagram doesn't match that, and I believe the text is correct. Therefore I think whoever drew the diagram made a mistake.

**Reply by jffaust** [reply]
What about gamma? I can't find any reference to it in the text. 

I see in the code the following formula for calculating the spin:

https://github.com/cosinekitty/astronomy/blob/master/source/js/astronomy.js#L8741

Was it taken from the [IERS](https://www.iers.org/)?

**Reply by cosinekitty** [reply]
I think the $\gamma$ in the diagram is intended to designate the equinox point. So it is a point label, not a numeric variable. Which is confusing when they go and add something to it! Because the equinox point has right ascension = 0, I just ignore it in the formula for calculating $Q$.

The formula for the Earth's rotation is adapted from [NOVAS C 3.1](https://aa.usno.navy.mil/software/novasc_intro). I have a copy of NOVAS checked into the repo because it is used for testing Astronomy Engine. Take a look at the [Earth Rotation Angle function `era`](https://github.com/cosinekitty/astronomy/blob/8fa81d0fbc26a709ac364059243b88e928e28dcb/generate/novas/novas.c#L3176). This leads to the [Astronomy Engine version of `era`](https://github.com/cosinekitty/astronomy/blob/8fa81d0fbc26a709ac364059243b88e928e28dcb/source/js/astronomy.js#L1836), which in turn leads to the modified version you linked whose value always increases with time, instead of cycling between 0..360 degrees. I also had to adjust its initial value to reflect the "expected" value at the J2000 epoch. But that begs the question: what is expected then? Unfortunately I'm having trouble finding/remembering how I came up with that. I'll report back here if/when I find a citation.

**Reply by cosinekitty** [reply]
OK, I found it! In the [2009 version](https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2009reprint.pdf) of the IAU Working Group Report, take a look at PDF page 7. Table 1 gives this very simple formula:

W = 190.147 + 360.9856235d

This doesn't look the same as what's in my code, but it actually works the same. The `d` in the formula is measured in atomic time, not civil time, in which the J2000 epoch is 2000-01-01T11:58:56.153Z. I use `ut` in my time structure which corrects for irregularities in the Earth's rotation over time. I will explain more later...

**Reply by cosinekitty** [reply]
Continuing where I left off, the simple formula in the 2009 report was abandoned in the 2015 report because it doesn't handle the fact that the Earth's rotation is slowing down over time. The `era` functions in NOVAS and Astronomy Engine do account for this by using UT (so-called Universal Time) instead of TT (Terrestrial Time).

The names UT and TT are unfortunately more confusing than helpful. It's better to think of UT as time defined by the rotation of the Earth, and TT as time measured by counting seconds from an atomic clock. TT keeps moving at a fixed rate, while UT keeps slowing down gradually over time as the Earth loses angular momentum. 

So that 2009 IAU formula works well near the J2000 epoch, but as you move forward or backward in time away from the year 2000, it gets more and more inaccurate.

So what I did was to take the single moment in time when the 2009 formula was most accurate, which is when d=TT=0, giving the value W = 190.147&deg;.  Then I needed to find the value of UT such that TT=0. It's easiest to show this using the Python version of Astronomy Engine. First I calculate the time value for TT=0:

```
>>> import astronomy
>>> d = astronomy.Time.FromTerrestrialTime(0)
>>> d
Time('2000-01-01T11:58:56.153Z')
>>> d.ut
-0.0007389709361535305
>>> d.tt
0
```

You can see that TT=0, but UT=-0.0007389709361535305 days. Then I calculate the NOVAS Earth Rotation Angle at this UT:

```
>>> a = astronomy._era(d)
>>> a
280.19386049918955
```

If I replace the `era` function with a simple linear function that is not wrapped to 0..360 degrees, it can be simplified as:

```
>>> def e(ut):
...     return 280.46061837504 + 360.9856122880876*ut
>>> e(d.ut)
280.1938604991895
```
Getting the same value (to 15 decimal digits) as `era` confirms it is working.  I need a function like `e`, only it should return 190.147 at the same moment of time. Therefore, I need to adjust the constant term downward by the discrepancy:

```
>>> 280.46061837504 - (e(d.ut) - 190.147)
190.41375787585048
```

This is the same constant term as I have in my code for `EarthRotationAxis`, at least to 10 decimal places. There is a tiny discrepancy here which I don't understand, but it is less than one ten millionth of a degree, which is not physically measurable, so I'm not going to worry about it.

Anyway, to verify consistency with the IAU 2009 formula:

```
>>> astronomy.RotationAxis(astronomy.Body.Earth, d)
AxisInfo(ra=15.077070372867658, dec=89.99777787194692, spin=190.14700001115204, north=Vector(-2.6865162337053374e-05, -2.797175571289667e-05, 0.9999999992479219, Time('2000-01-01T11:58:56.153Z')))
```

And you can see the spin angle is very close to 190.147&deg; as required.

The summary is, the angle starts out the same as defined in their document, only I corrected it to account for the Earth's rotation changing over time. That rotation slowdown is handled by the `DeltaT_EspenakMeeus` function, in case you're interested. There are citation links in the source code to better understand how the &Delta;T function works.

**Reply by jffaust** [reply]
Thank you for the detailed explanation. It helps.

I understand the formula for $W$ and how it considers that the Earth's rotation is slowing over time. What I don't understand is the logic of defining the node $Q$. From the 2015 report, the text and illustration mention using the body's equator of date. If I am getting the spin or $W$ from the engine for various dates, doesn't the body's equator of date change over time too, thus shifting the reference point $Q$? 

The formula for calculating the spin is relatively simple and in my head it doesn't imply a potentially shifting starting point. It's just  a starting angle plus a number of degrees rotated for a given fractional number of UT. It just spins, relative to its rotational axis and it doesn't really matter if that axis moves because of precession and nutation. 

Am I missing something?

**Reply by cosinekitty** [reply]
No, you aren't missing anything. That is a perceptive observation! I don't think that really clicked with me until you pointed it out.

An excerpt from the 2015 report (page 6 of 46, section 2), with some emphases added by me:

> As long as the planet rotates **uniformly**,
W varies **nearly** linearly with time. The parameters **α0, δ0, and W may vary with time** due to
a precession of the axis of rotation of the planet or satellite.

So even with uniform rotation, $W$ varies "nearly" linearly, not exactly!

The first paragraph in that section makes a concession:

> **Approximate expressions** for these rotational elements with respect to the International
Celestial Reference Frame ...  have been derived.

I feel a little bit like a lawyer, but putting these two excerpts together makes me think this is why they removed the Earth rotation axis model that was present in the 2009 paper. The simple linear formula probably annoyed a lot of scientists who knew it was **too** simple, especially compared with extra terms they gave for other planets.

For example, in their formula for Mars, the calculation of $\alpha_0$, $\delta_0$, and $W$ are much more complex. My understanding is that the extra trigonometric terms reflect precession of Mars' axis with respect to the ICRF equator.

As goofy as it sounds, they really do measure $W$ with respect to the shifting intersection point between the body's equator of date and the ICRF plane, not with respect to the total angular rotation around the planet's own axis. I think there are practical pros and cons to both ways of defining rotation, and this IAU working group chose the one that allows calculation of the body at any moment in time without needing to include its state at the epoch in each calculation.

So this makes me realize that over time, my EarthRotationAxis formula for $W$ is also deficient. It's better than the IAU 2009 formula in that it compensates for the Earth's spin slowing down, but it doesn't handle the ever-shifting intersection nodes between equator of date and the ICRF equator.

So, I'm thinking maybe it's a good time to step back and revisit your original goal. If I were to make an accurate 3D rendering of the Solar System, I would be tempted to use a different approach for depicting the Earth that I know has a stronger footing. For each moment of time in your simulation, consider calling the function [ObserverVector](https://github.com/cosinekitty/astronomy/tree/master/source/js#Rotation_EQD_ECL) three times, once for 3 different observers:

latitude=0, longitude=0, height=0
latitude=0, longitude=+90, height=0
latitude=90, longitude=0, height=0

Pass in `true` for the `ofdate` parameter. This will give you 3 points on the rotating oblate sea-level geoid pointing in 3 mutually perpendicular directions, defined in the equator-of-date system.

Then call [Rotation_EQD_ECL](https://github.com/cosinekitty/astronomy/tree/master/source/js#Rotation_EQD_ECL) to obtain a rotation matrix that converts equator-of-date coordinates to ecliptic coordinates, since that is your preferred orientation system. Use [RotateVector](https://github.com/cosinekitty/astronomy/tree/master/source/js#RotateVector) to convert each of the 3 equatorial vectors into ecliptic vectors.

At this point you have a complete representation of the orientation of the Earth's oblate shape in space, along with the rotated orientation of continents and oceans. I believe this will side-step any confusion of the IAU way of looking at rotation.

Does this help?

**Reply by jffaust** [reply]
"oblate", "geoid" and I thought I was getting closer to an accurate representation. Back to internet research ahah. 

Thank you again for the response. I will report back later :)

**Reply by cosinekitty** [reply]
If it helps, Astronomy Engine uses the [WGS 84 model](https://en.wikipedia.org/wiki/World_Geodetic_System) of the Earth's shape. The tl;dr version is, the Earth's equatorial radius is 6378.1366 km and its polar radius is 6356.7519 km.

You might get away with not depicting the Earth as oblate because the difference is pretty small. However, you really can't do Jupiter justice without showing its pronounced rotational bulge. The Sun spins so slowly that it is a perfect sphere for all practical purposes. Same thing with Venus.

The main gotcha that often confuses people in WGS 84 is the difference between geographic (a.k.a. geodetic) latitude and geocentric latitude. Geographic latitude, as used for maps and GPS, is based on measuring star positions from the observer's horizon, while geocentric latitude is based on the angle between two lines: a line from the center of the Earth passing through the observer, and the line from the center of the Earth passing through the equator due north/south of the observer.

It turns out that if you are standing somewhere between the equator and the north pole, the line that passes straight down, perpendicular to your observed sea-level horizon, will not pass through the center of the Earth, because the Earth is spinning and thus oblate. The sea level surface has two accelerations acting on it: gravity and centripetal acceleration, and they don't point along the same line. So the ocean doesn't flatten perpendicular to the center of the Earth. Hopefully the distinction doesn't matter at the resolution you are graphing, but if it does, let me know and I can help.

**Reply by jffaust** [reply]
I think it works now! I will do some more validation and at some point share the results :)

Thanks so much for your help :D
